python/daco2ego.py

```diff
@@ -104,21 +114,35 @@ def init(config):
     client_id = config['client']['client_id']
     client_secret = config['client']['client_secret']
     base_url = config['client']['base_url']
+    dac_api_url = config['client']['dac_api_url']
 
     rest_client = get_oauth_authenticated_client(base_url, client_id, client_secret)
 
-    ego_client = EgoClient(base_url, rest_client,  # Want to create a factory for new oauth clients
+    ego_client = EgoClient(base_url, rest_client, dac_api_url,  # Want to create a factory for new oauth clients
                            lambda: get_oauth_authenticated_client(base_url, client_id, client_secret))
 
     encoding_override = config.get('file_encoding_override', None)
 
     daco = csv_to_dict(decrypt_file(config['daco_file'], key, iv, hexdump=hexdump), encoding_override)
     cloud = csv_to_dict(decrypt_file(config['cloud_file'], key, iv, hexdump=hexdump), encoding_override)
-    users = get_users(daco, cloud)
+    usersFromDacApi = ego_client.download_daco2_approved_users()
+
+    daco1_users = get_users(daco, cloud)
+    daco2_users = daco2_csv_to_dict(usersFromDacApi)
+
+    # concatenate daco1 users from csv and daco2 users from dac-api endpoint
+    users = daco1_users + daco2_users
+
+    # sort users by has_cloud key as daco2 users will always have has_cloud=True but daco1 users can have False
+    # if a duplicate user does not have cloud access in daco1 but does have it in daco2, we want to use the daco2 access
+    sorted_users = sorted(users, key=lambda k: k.has_cloud, reverse=False)
+
+    # create a temporary ordered dict on email key to filter out duplicates
+    deduplicated_users = list(OrderedDict((v.email, v) for v in sorted_users).values())
```

Comment:
I'm not sure I follow this.

Based on this precondition:
```
    # sort users by has_cloud key as daco2 users will always have has_cloud=True but daco1 users can have False
    # if a duplicate user does not have cloud access in daco1 but does have it in daco2, we want to use the daco2 access
```

Does it not make more sense to construct a dict first from `daco2` which are guarnateed to have `cloud=true` and then only conditionally add users from `daco1` if they are not already in this dict?

if both lists are approximately size `n`, then the worst case for this is `O(n)`.  Specifically `2n` operations.

If I'm understanding this code, you construct the full list `2n operations`, then sort them `nlog(n)` and then construct an ordered dict, and then from that a list. 

Maybe I'm missing something fundamental about the usage of the ordered dict here.

Reply:
No I think you're right - i was struggling with how to handle merging the two lists, i did get as far as knowing i should at least *start* with the daco2 list as the main source, but after that it got a little hazy. The approach with the ordereddict was to make sure of the sort order, so a daco2 user would be added to the final list first, ensuring that if they existed in daco1 without cloud, that wouldn't override their daco2 access.
But it's definitely not a clear solution. For your suggestion, do you mean to still call the old `csv_to_dict` on daco1 users (so every item is a `User`), then map through that list to add to the daco2 list? Or instead inside `csv_to_dict`, check the daco2 list and only add users that don't yet exist in the list?

Reply:
this is not so much a `map` rather than a `forEach` type of op.

I'm not sure how this is being marshalled as I don't have the type information, but I'm assuming both lists are made up of some sort of User object, be it a typed python class or some json. 

So you build a dict of <email, user_obj> from daco2. Then you iterate through the list of users in daco1. For each user in daco1, you can constant time lookup if it exists in your dict based on email as key. If it does, do nothing (since it will already have the superset of cloud=true. If it does not exist, add it do this dict which is also constant time.  At the end if you simply want a list rather than a dict, just call `.values()`. 

No need for an ordered dict or sorting.